packages/contracts/src/utils/RelayRegistrar.sol

```diff
@@ -9,153 +9,124 @@ import "../interfaces/IRelayHub.sol";
         emit RelayServerRegistered(relayManager, baseRelayFee, pctRelayFee, url);
         if (isUsingStorageRegistry) {
-            storeRelayServerRegistration(relayManager, baseRelayFee, pctRelayFee, url);
+            storeRelayServerRegistration(relayHub, relayManager, baseRelayFee, pctRelayFee, url);
         }
     }
 
-    function addItem(address relayManager) internal returns (RelayStorageInfo storage) {
-        RelayStorageInfo storage storageInfo = values[relayManager];
-        if (storageInfo.lastBlockNumber == 0) {
-            indexedValues.push(relayManager);
+    function addItem(address relayHub, address relayManager) internal returns (RelayInfo storage) {
+        RelayInfo storage storageInfo = values[relayHub][relayManager];
+        if (storageInfo.lastSeenBlockNumber == 0) {
+            indexedValues[relayHub].push(relayManager);
         }
         return storageInfo;
     }
 
-    function storeRelayServerRegistration(address relayManager, uint256 baseRelayFee, uint256 pctRelayFee, string calldata url) internal {
-        RelayStorageInfo storage storageInfo = addItem(relayManager);
-        if (storageInfo.stakeBlockNumber==0) {
-            storageInfo.stakeBlockNumber = uint32(block.number);
+    function storeRelayServerRegistration(
+        address relayHub,
+        address relayManager,
+        uint96 baseRelayFee,
+        uint96 pctRelayFee,
+        bytes32[3] calldata url
+    ) internal {
+        RelayInfo storage storageInfo = addItem(relayHub, relayManager);
+        if (storageInfo.firstSeenBlockNumber==0) {
+            storageInfo.firstSeenBlockNumber = uint32(block.number);
         }
-        storageInfo.lastBlockNumber = uint32(block.number);
-        storageInfo.baseRelayFee = uint96(baseRelayFee);
-        storageInfo.pctRelayFee = uint96(pctRelayFee);
-        bytes32[3] memory parts = splitString(url);
-        storageInfo.urlParts = parts;
+        storageInfo.lastSeenBlockNumber = uint32(block.number);
+        storageInfo.baseRelayFee = baseRelayFee;
+        storageInfo.pctRelayFee = pctRelayFee;
+        storageInfo.relayManager = relayManager;
+        storageInfo.urlParts = url;
     }
 
     /// @inheritdoc IRelayRegistrar
-    function getRelayInfo(address relayManager) public view override returns (RelayInfo memory) {
-        RelayInfo memory info;
-        RelayStorageInfo storage storageInfo = values[relayManager];
-        require(storageInfo.lastBlockNumber != 0, "relayManager not found");
-        info.lastBlockNumber = storageInfo.lastBlockNumber;
-        info.stakeBlockNumber = storageInfo.stakeBlockNumber;
-        info.baseRelayFee = storageInfo.baseRelayFee;
-        info.pctRelayFee = storageInfo.pctRelayFee;
-        info.relayManager = relayManager;
-        info.url = packString(storageInfo.urlParts);
+    function getRelayInfo(address relayHub, address relayManager) public view override returns (RelayInfo memory) {
```

Comment:
change return value to "RelayInfo memory info", and thus copying `info` into return value

Reply:
I am not sure this is how it works. Defining a variable name places it in the returndata memory? How do you know that?